File: Event/Stock.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Event
{
    public class Stock
    {
        //public delegate void OnChangedPrice(string message);
        //public event OnChangedPrice onChangedPrice;

        //public event Action<string> onChangedPrice;

        public event EventHandler<MessageEventArg> ChangedPrice;//استاندارد تعریف ایونت باید مشابه این باشد

        protected virtual void OnChangedPrice(MessageEventArg e)
        {
            ChangedPrice?.Invoke(this, e);
        }
        public Stock()
        {

        }
        public decimal Price { get; set; }
        public void UpdatePrice(decimal price)
        {
            Price = price;
            OnChangedPrice( new MessageEventArg($"Price Update To {price}"));
        }
    }

    public class MessageEventArg(string message) : EventArgs
    {
        public string Message =message;
    }
}

File: Event_Vb/Stock.vb
Public Class Stock

    'Public Delegate Sub OnPriceChangedHandler(ByVal message As String)
    'Public Event OnPriceChanged As OnPriceChangedHandler
    Public Event PriceChanged As EventHandler(Of MessageEventArgs)
    Protected Overridable Sub OnPriceChanged(ByVal e As MessageEventArgs)
        RaiseEvent PriceChanged(Me, New MessageEventArgs($"new price is : {Price}"))
    End Sub
    Public Property Price As Integer
    Public Sub Update(ByVal newPrice As Integer)
        Price = newPrice
        OnPriceChanged(New MessageEventArgs($"new price is : {Price}"))
    End Sub
End Class

Public Class MessageEventArgs
    Inherits EventArgs
    Public ReadOnly Property Text As String
    Sub New(ByVal text As String)
        Me.Text = text
    End Sub
End Class
